gateway/handler_unixfs__redirects.go

```diff
@@ -36,46 +38,59 @@ import (
 //
 // Note that for security reasons, redirect rules are only processed when the request has origin isolation.
 // See https://github.com/ipfs/specs/pull/290 for more information.
-func (i *handler) serveRedirectsIfPresent(w http.ResponseWriter, r *http.Request, resolvedPath ipath.Resolved, contentPath ipath.Path, logger *zap.SugaredLogger) (newResolvedPath ipath.Resolved, newContentPath ipath.Path, continueProcessing bool, hadMatchingRule bool) {
-	redirectsFile := i.getRedirectsFile(r, contentPath, logger)
-	if redirectsFile != nil {
-		redirectRules, err := i.getRedirectRules(r, redirectsFile)
-		if err != nil {
-			webError(w, err, http.StatusInternalServerError)
-			return nil, nil, false, true
-		}
+func (i *handler) serveRedirectsIfPresent(w http.ResponseWriter, r *http.Request, maybeResolvedImPath, immutableContentPath ImmutablePath, contentPath ipath.Path, logger *zap.SugaredLogger) (newContentPath ImmutablePath, continueProcessing bool, hadMatchingRule bool) {
+	// contentPath is the full ipfs path to the requested resource,
+	// regardless of whether path or subdomain resolution is used.
+	rootPath := getRootPath(maybeResolvedImPath)
```

Comment:
:warning: :thought_balloon:  Why do we use `maybeResolvedImPath` here and not `immutableContentPath`?

This feels like a potential bug when 404  happens on deeper paths:   we would be looking for `_redirects` in the wrong directory, no?

- Let's assume contentPath: `/ipns/example.net/foo/bar` and the `bar` being not found (404)
- :green_circle:  immutableContentPath: `getRootPath(/ipfs/cid/foo/bar/)` → we check `/ipfs/cid/_redirects`  (ok)
-  :red_circle:  maybeResolvedImPath: `getRootPath(/ipfs/cid-of-bar/)` → `/ipfs/cid/foo/bar/_redirects` (invalid, we only look for _redirects in the top level root directory)


```suggestion
	rootPath := getRootPath(immutableContentPath)
```

If sharness did not catch this, we are probably testing only one level.

Reply:
I think `maybeResolvedImPath != immutableContentPath` if the header `If-None-Match` is set and is valid. I don't mind changing this, but I would like to have a reproducible way of testing this.

I'm trying to understand why we need the `maybeResolvedImPath` at all. Before we were using `resolvedPath` in all handlers, which I now think is `immutableContentPath`? Why do we have two now? And why was `resolvedPath` replaced by `maybeResolvedImPath` and not `immutableContentPath` in the other handlers?

Reply:
I'd like to simplify/clean this up, even if we only rename variables or add new type.

## In old version..

In the old version, we had very clear model:
- `contentPath` is the original content path what was requested by the user (`/ipfs/cid/sub/file` or `/ipns/example.com/sub/file`
- `resolvedPath` was `ipfs resolve -r $contentPath` – always final `/ipfs/cid` + optional remainder if the path was for a field in CBOR document

It is an important distinction, because we have to use unresolved, original path to make some decisions.
For example, things like `Cache-Control` on `/ipns/` content path will be different than `/ipfs/`, same with searching for `_redirects` only at the root of a website.  

### This PR.. 

In this PR, things got fuzzy and that creates space for regressions (especially for gateway behaviors before my time).

`immutableContentPath`  is different from the old `resolvedPath`  because it is not resolved to the final `/ipfs/cidCID, it only resolves `/ipns/id` to `/ipfs/cid`.   

I have no idea how to interpret `maybeResolvedImPath`. 

@aschmahmann is the only reason why we have `maybeResolvedImPath` is to avoid additional roundtrip after we switch to partial CAR requests in biforst-gateway?   Is it resolved or not? We can't have "maybe", because it creates surface for  bugs around things like `_redirects`, probably more in the future.

Reply:
The old code called `api.ResolvePath` basically all the time assuming that all the data was cached locally and it wouldn't be a costly resolution.

`maybeResolvedImPath` takes into account that you might have already done the path resolution as a function of `If-None-Matched` requiring it and therefore not wanting to do the lookup again.

Generally speaking part of the refactor here is to reduce the assumptions on the underlying layer being kubo behaving as kubo does today (i.e. block requests go into a blockstore that get cleaned up on GC) which includes not doing extra lookups.

IIRC the main remaining source of extra lookups is in the legacy 404 code which could be resolved, but I didn't put much effort into it given that it's "legacy".

Yes, it looks like `getRootPath` should work on the immutable content path without resolution. Good catch 🙏

Reply:
> is the only reason why we have maybeResolvedImPath is to avoid additional roundtrip after we switch to partial CAR requests in biforst-gateway? Is it resolved or not?

It's not so much about round-trips as about extra work in general (without implementation specifics we can't really comment about what the form of that work will be). For example, if I do an `If-None-Match` for `/ipfs/bafyfoo/dir1/dir2/dir3/dir4/file` and it's not a match and I'm not using `maybeResolvedImPath` then I have to do the path resolution again. If all the blocks are local, then I have some disk and CPU usage but not too bad. If the blocks are no longer local (e.g. my gateway implementation threw away the blocks after it fulfilled the request) then it's another network request. If that network request is backed by block fetching now it's a whole bunch of round-trips.

Reply:
This has been now fixed.